File: etl/common/timeofday.py
"""Modeled (estimated) time-of-day layer.

The FIR extract has NO clock time (only Year/Month/Day), so true time-of-day is
unrecoverable. We construct an *estimated* time-of-day bucket per incident,
conditioned on crime type, using two clearly-separated methods:

  1. ``category_encoded`` -- REAL signal already embedded in the crime classification.
     Some major heads encode day/night directly, e.g. "BURGLARY - NIGHT" /
     "BURGLARY - DAY", "HOUSE BREAKING BY NIGHT". Used verbatim.
  2. ``criminological_prior`` -- documented per-category priors for the rest
     (e.g. road accidents skew to evening/commute, riots/affray to evening,
     robbery/dacoity to night). These are illustrative assumptions, not observations.
  3. ``default_distributed`` -- no reliable prior -> "Distributed".

HONESTY: every output carries data_class="modeled" and the method used. This layer is
for the illustrative time x location view ONLY. It is NEVER used to train real models.

Buckets: Night, Daytime, Morning, Afternoon, Evening, Distributed.
"""
from __future__ import annotations

BUCKETS = ["Night", "Daytime", "Morning", "Afternoon", "Evening", "Distributed"]

# Keyword (substring, matched against the cleaned UPPER major head) -> bucket.
# Order matters: first match wins. Rationale noted inline.
_PRIORS = [
    # --- REAL signal encoded in the category (method = category_encoded) ---
    ("BURGLARY - NIGHT", "Night", "category_encoded"),
    ("BURGLARY - DAY", "Daytime", "category_encoded"),
    ("BY NIGHT", "Night", "category_encoded"),
    ("BY DAY", "Daytime", "category_encoded"),
    # --- criminological priors (method = criminological_prior) ---
    ("ROBBERY", "Night", "criminological_prior"),        # street robbery skews night
    ("DACOITY", "Night", "criminological_prior"),
    ("BURGLARY", "Night", "criminological_prior"),       # unspecified burglary -> night lean
    ("HOUSE BREAK", "Night", "criminological_prior"),
    ("MURDER", "Night", "criminological_prior"),          # incl. ATTEMPT TO MURDER
    ("MOTOR VEHICLE ACCIDENT", "Evening", "criminological_prior"),  # commute/night driving
    ("NEGLIGENT", "Evening", "criminological_prior"),
    ("RIOT", "Evening", "criminological_prior"),
    ("AFFRAY", "Evening", "criminological_prior"),
    ("UNLAWFUL ASSEMBLY", "Evening", "criminological_prior"),
    ("PUBLIC SAFETY", "Evening", "criminological_prior"),
    ("PUBLIC NUISANCE", "Evening", "criminological_prior"),
    ("NARCOTIC", "Night", "criminological_prior"),
    ("EXCISE", "Evening", "criminological_prior"),
    ("KARNATAKA POLICE ACT", "Evening", "criminological_prior"),  # gambling/liquor/order
    ("GAMBLING", "Evening", "criminological_prior"),
    ("COTPA", "Evening", "criminological_prior"),
    ("IMMORAL TRAFFIC", "Night", "criminological_prior"),
    ("THEFT", "Afternoon", "criminological_prior"),       # opportunistic, daytime lean
    ("TRESPASS", "Afternoon", "criminological_prior"),
]

_DEFAULT = ("Distributed", "default_distributed")


def assign(major_head_clean: str):
    """Return (bucket, method) for a cleaned major head string."""
    h = (major_head_clean or "").upper()
    if not h:
        return _DEFAULT[0], _DEFAULT[1]
    for kw, bucket, method in _PRIORS:
        if kw in h:
            return bucket, method
    return _DEFAULT[0], _DEFAULT[1]

File: functions/crime_api/src/routes/timeofday.js
"use strict";
const { getTable } = require("../lib/store");
const { loadDistrictIndex, num } = require("../lib/districts");

// Bucket display order (Phase 0 emits: Night, Daytime, Morning, Afternoon, Evening, Distributed)
const ORDER = ["Night", "Morning", "Daytime", "Afternoon", "Evening", "Distributed"];
const NOTE =
  "Estimated time-of-day — MODELLED, not observed. The FIR extract has no clock time; " +
  "buckets are derived from the crime head's day/night signal + documented criminological " +
  "priors. Illustrative for the time×location view only; never used to train models.";

/**
 * GET /timeofday?district=&category=  — MODELLED time-of-day distribution (data_class=modeled).
 */
module.exports = (router, asyncH) => {
  router.get("/timeofday", asyncH(async (req, res) => {
    const rows = await getTable("agg_timeofday", req.ctx);
    const d = req.query.district ? decodeURIComponent(req.query.district) : null;
    const cat = req.query.category ? decodeURIComponent(req.query.category) : null;

    let memberSet = null;
    if (d) {
      const idx = await loadDistrictIndex(req.ctx);
      const members = idx.membersByParent[d] || (idx.parentByCanonical[d] ? idx.membersByParent[idx.parentByCanonical[d]] : null);
      if (members) memberSet = new Set(members);
    }

    const buckets = new Map();
    const methods = new Map();
    let total = 0;
    for (const r of rows) {
      if (memberSet && !memberSet.has(r.canonical_name)) continue;
      if (cat && r.major_head !== cat) continue;
      const c = num(r.count);
      total += c;
      buckets.set(r.modeled_time_of_day, (buckets.get(r.modeled_time_of_day) || 0) + c);
      methods.set(r.method, (methods.get(r.method) || 0) + c);
    }

    const bucketArr = [...buckets.entries()]
      .map(([bucket, count]) => ({ bucket, count }))
      .sort((a, b) => {
        const ia = ORDER.indexOf(a.bucket), ib = ORDER.indexOf(b.bucket);
        return (ia < 0 ? 99 : ia) - (ib < 0 ? 99 : ib) || b.count - a.count;
      });
    const methodArr = [...methods.entries()].map(([method, count]) => ({ method, count })).sort((a, b) => b.count - a.count);

    res.sendOk({ district: d, category: cat, total, buckets: bucketArr, by_method: methodArr, note: NOTE }, "modeled");
  }));
};
